backend/src/org/commcare/session/CommCareSession.java

```diff
@@ -370,6 +394,26 @@ public void setDatum(String keyId, String value) {
         syncState();
     }
 
+    /**
+     * Set a (xml) data instance as the result to a session query datum.
+     * The instance is available in session's evaluation context until the corresponding query frame is removed
+     */
+    public void setQueryDatum(ExternalDataInstance queryResultInstance) {
+        SessionDatum datum = getNeededDatum();
+        if (datum instanceof RemoteQueryDatum) {
+            StackFrameStep step =
+                    new StackFrameStep(SessionFrame.STATE_QUERY_REQUEST,
+                            datum.getDataId(), datum.getValue());
+            frame.pushStep(step);
```

Comment:
@phillipm I agree with this; it seems a lot more well-contained to make the instance a part of the `StackFrameStep` itself, instead of relying on keeping this random `instances` member variable in sync with the stack.